data/syncer/userAccountsSyncer.go

```diff
@@ -0,0 +1,164 @@
+	WaitTime           time.Duration
+	ShardId            uint32
+	Cacher             storage.Cacher
+}
+
+// NewUserAccountsSyncer creates a user account syncer
+func NewUserAccountsSyncer(args ArgsNewUserAccountsSyncer) (*userAccountsSyncer, error) {
+	if check.IfNil(args.Hasher) {
+		return nil, state.ErrNilHasher
+	}
+	if check.IfNil(args.Marshalizer) {
+		return nil, state.ErrNilMarshalizer
+	}
+	if check.IfNil(args.TrieStorageManager) {
+		return nil, state.ErrNilStorageManager
+	}
+	if check.IfNil(args.RequestHandler) {
+		return nil, state.ErrNilRequestHandler
+	}
+	if args.WaitTime < time.Second {
+		return nil, state.ErrInvalidWaitTime
+	}
+
+	u := &userAccountsSyncer{
+		hasher:             args.Hasher,
+		marshalizer:        args.Marshalizer,
+		trieSyncers:        make(map[string]data.TrieSyncer),
+		dataTries:          make(map[string]data.Trie),
+		trieStorageManager: args.TrieStorageManager,
+		requestHandler:     args.RequestHandler,
+		waitTime:           args.WaitTime,
+		shardId:            args.ShardId,
+		cacher:             args.Cacher,
+		rootHash:           nil,
+	}
+
+	return u, nil
+}
+
+// SyncAccounts will launch the syncing method to gather all the data needed for userAccounts
+func (u *userAccountsSyncer) SyncAccounts(rootHash []byte) error {
+	u.rootHash = rootHash
+
+	dataTrie, err := trie.NewTrie(u.trieStorageManager, u.marshalizer, u.hasher)
+	if err != nil {
+		return err
+	}
+
+	u.dataTries[string(rootHash)] = dataTrie
+	trieSyncer, err := trie.NewTrieSyncer(u.requestHandler, u.cacher, dataTrie, u.waitTime, u.shardId, factory.AccountTrieNodesTopic)
+	if err != nil {
+		return err
+	}
+	u.trieSyncers[string(rootHash)] = trieSyncer
+
+	err = trieSyncer.StartSyncing(rootHash)
+	if err != nil {
+		return err
+	}
+
+	rootHashes, err := u.findAllAccountRootHashes(dataTrie)
+	if err != nil {
+		return err
+	}
+
+	err = u.syncAccountDataTries(rootHashes)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (u *userAccountsSyncer) syncAccountDataTries(rootHashes [][]byte) error {
+	for _, rootHash := range rootHashes {
+		dataTrie, err := trie.NewTrie(u.trieStorageManager, u.marshalizer, u.hasher)
+		if err != nil {
+			return err
+		}
+
+		u.dataTries[string(rootHash)] = dataTrie
+		trieSyncer, err := trie.NewTrieSyncer(u.requestHandler, u.cacher, dataTrie, u.waitTime, u.shardId, factory.AccountTrieNodesTopic)
+		if err != nil {
+			return err
+		}
+		u.trieSyncers[string(rootHash)] = trieSyncer
+
+		err = trieSyncer.StartSyncing(rootHash)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (u *userAccountsSyncer) findAllAccountRootHashes(mainTrie data.Trie) ([][]byte, error) {
+	leafs, err := mainTrie.GetAllLeaves()
+	if err != nil {
+		return nil, err
+	}
+
+	rootHashes := make([][]byte, 0)
+	for _, leaf := range leafs {
+		account := &state.Account{}
+		err := u.marshalizer.Unmarshal(account, leaf)
+		if err != nil {
+			log.Trace("this must be a leaf with code", "err", err)
```

Comment:
Code is saved as a []byte in the leaf. When accessing the leaf there is no information whether it is a code / account. So first there is a try to unmarshal into an account, otherwise it must be code.